File: python/system/gpr/variables/sources.py
from numpy import exp, inf

from system.gpr.misc.functions import det3, sigma_norm
from system.gpr.variables.state import sigma


def theta1inv(A, PAR):
    """ Returns 1/θ1
    """
    cs2 = PAR.cs2
    τ1 = PAR.τ1

    if τ1==inf:
        return 0

    if PAR.PLASTIC:
        σY = PAR.σY
        n = PAR.n
        ρ = det3(A) * PAR.ρ0
        σ = sigma(ρ, A, PAR)
        sn = sigma_norm(σ)
        if sn == 0:
            return 0
        τ = τ1 * (σY / sn) ** n
    else:
        τ = τ1

    return 3 * det3(A)**(5/3) / (cs2 * τ)

def theta2inv(ρ, T, PAR):
    """ Returns 1/θ2
    """
    ρ0 = PAR.ρ0
    T0 = PAR.T0
    α2 = PAR.α2
    τ2 = PAR.τ2

    return 1 / (α2 * τ2 * (ρ / ρ0) * (T0 / T))

def K_arr(P, PAR):
    """ Returns the rate of reaction according to Arrhenius kinetics
    """
    ρ = P.ρ
    λ = P.λ
    T = P.T

    Bc = PAR.Bc
    Ea = PAR.Ea
    Rc = PAR.Rc

    return Bc * ρ * λ * exp(-Ea / (Rc*T))

def K_dis(P, PAR):
    """ Returns the rate of reaction according to discrete ignition temperature
        reaction kinetics
    """
    ρ = P.ρ
    λ = P.λ
    T = P.T

    Ti = PAR.Ti
    Kc = PAR.Kc

    if T > Ti:
        return ρ * λ * Kc
    else:
        return 0

def K_ing(P, PAR):
    """ Returns the rate of reaction according to ignition and growth
        reaction kinetics
    """
    I = PAR.I
    G1 = PAR.G1
    G2 = PAR.G2
    a = PAR.a
    b = PAR.b
    c = PAR.c
    d = PAR.d
    e = PAR.e
    g = PAR.g
    x = PAR.x
    y = PAR.y
    z = PAR.z
    φIG = PAR.φIG
    φG1 = PAR.φG1
    φG2 = PAR.φG2

    ρ = P.ρ
    p = P.p
    λ = P.λ
    φ = 1 - λ

    ret = 0
    if φIG > φ:
        ret += I * λ**b * (ρ-1-a)**x
    if φG1 > φ:
        ret += G1 * λ**c * φ**d * p**y
    if φ > φG2:
        ret += G2 * λ**e * φ**g * p**z

    return ret

File: code/cpp/src/system/variables/sources.cpp
#include <cmath>

#include "../../etc/globals.h"
#include "../functions/matrices.h"
#include "../functions/vectors.h"
#include "../variables/state.h"
#include "../waves/shear.h"

const int DISCRETE = 0;
const int ARRHENIUS = 1;
const int IGNITION_GROWTH = 2;

double theta1inv(VecVr Q, Par &MP) {
  // Returns the relaxation parameter for the distortion tensor

  // TODO: implement mixing of μ

  double ρ = Q(0);
  Mat3_3Map A = get_A(Q);

  double A53 = pow(A.determinant(), 5. / 3.);
  double cs2 = c_s2(ρ, MP);
  double n = MP.n;

  double sn;
  if (MP.SOLID) {

    if (MP.POWER_LAW) {

      Mat3_3 σ = sigma(Q, MP);
      sn = sigma_norm(σ);
      sn = std::min(sn, 1e8); // Hacky fix

      double σY = MP.σY;
      return 3 * A53 / (cs2 * MP.τ0) * pow((sn / σY), n);
    } else
      return 0.;
  } else if (MP.BINGHAM) {

    Mat3_3 σ = sigma(Q, MP);
    sn = σ.norm() / sqrt(2.);
    double σY = MP.σY;
    if (sn > σY)
      return A53 * MP.ρ0 / (2 * MP.μ) * (1. - σY / sn);
    else
      return 0.;
    /*
    double τf = MP.τf;
    double τs = MP.τs;
    double bf2 = MP.bf2;
    double bs2 = MP.bs2;
    double ε = 1e-9;

    Mat3_3 σ = sigma(Q, MP);
    double Γ = sigma_norm(σ) / MP.σY;
    // double Γ = σ.norm() / sqrt(2.) / MP.σY;
    double h = 0.5 * (1. + Γ - sqrt(1. + (4. * ε - 2.) * Γ + Γ * Γ));

    double λ;
    if (Γ < ε)
      λ = 1.;
    else
      λ = h / Γ;
    λ = std::min(λ, 1.);

    double τ = τf * τs / (λ * τf + (1 - λ) * τs);
    double Cs = λ * sqrt(bs2) + (1 - λ) * sqrt(bf2);
    return 3. * A53 / (Cs * Cs * τ);
    */
  } else {
    double c = A53 * MP.ρ0 / (2 * std::pow(MP.μ, 1 / n));

    if (MP.POWER_LAW) {
      Mat3_3 σ = sigma(Q, MP);
      sn = σ.norm() / sqrt(2.);

      return c * pow(sn, (1 - n) / n);
    } else
      return c;
  }
}

double theta2inv(VecVr Q, Par &MP) {
  // Returns the relaxation parameter for the thermal impulse vector

  // TODO: implement mixing of κ

  double ρ = Q(0);
  double T = temperature(Q, MP);
  return T / (MP.κ * ρ);
}

void f_body(Vec3r x, Par &MP) { x = MP.δp; }

double reaction_rate(VecVr Q, Par &MP) {

  double λ = Q(mV) / Q(0);
  λ = std::max(0., λ);

  switch (MP.REACTION) {

  case DISCRETE: {
    if (temperature(Q, MP) > MP.Ti)
      return MP.Kc * λ;
    else
      return 0.;
  }
  case ARRHENIUS: {
    double T = temperature(Q, MP);
    return MP.Bc * exp(-MP.Ea / (MP.Rc * T)) * λ;
  }
  case IGNITION_GROWTH: {
    double ρ = Q(0);
    double p = pressure(Q, MP);
    p = std::max(p, 0.);
    λ = std::min(λ, 1.);
    double φ = 1 - λ;

    double ret = 0.;

    if (φ < MP.φI)
      ret += MP.I * std::pow(λ, MP.b) * std::pow(ρ / MP.ρ0 - 1 - MP.a, MP.x);

    if (φ < MP.φG1)
      ret += MP.G1 * std::pow(λ, MP.c) * std::pow(φ, MP.d) * std::pow(p, MP.y);

    if (φ > MP.φG2)
      ret += MP.G2 * std::pow(λ, MP.e) * std::pow(φ, MP.g) * std::pow(p, MP.z);

    return std::max(ret, 0.);
  }
  default:
    throw "Reaction type not recognized";
  }
}
